File: src/visual-graph/builder/push-edge.ts
import type { BuildState } from "./build-state.js";

export function pushEdge(
  state: BuildState,
  from: string,
  label: string,
  to: string,
): void {
  const key = `${from} -->|${label}| ${to}`;
  if (state.emittedEdges.has(key)) {
    return;
  }
  state.emittedEdges.add(key);
  state.edges.push({ from, to, label });
}

File: crates/unsnarl-visual-graph/src/builder/push_edge.rs
//! Deduplicates emitted edges by a `from -->|label| to` key so the
//! one-edge-per-(from, to, label) invariant is preserved across
//! every site that pushes into `state.edges`.

use std::collections::HashSet;

use crate::visual_edge::VisualEdge;

pub fn push_edge(
    emitted_edges: &mut HashSet<String>,
    edges: &mut Vec<VisualEdge>,
    from: &str,
    label: &str,
    to: &str,
) {
    let key = format!("{from} -->|{label}| {to}");
    if !emitted_edges.insert(key) {
        return;
    }
    edges.push(VisualEdge::new(from, to, label));
}

#[cfg(test)]
#[path = "push_edge_test.rs"]
mod push_edge_test;
